Add support for topOrigin and crossOrigin to [_WKWebAuthenticationPanel getClientDataJSONFor...]

abigailfox opened the pull request:
#### 8e0e561a0e0f49ddfe39a48d04c50ca1f6ea263e
<pre>
Add support for topOrigin and crossOrigin to [_WKWebAuthenticationPanel getClientDataJSONFor...]
<a href="https://bugs.webkit.org/show_bug.cgi?id=255680">https://bugs.webkit.org/show_bug.cgi?id=255680</a>
rdar://108064670

Reviewed by J Pascoe and Chris Dumez.

* Source/WebCore/Modules/webauthn/AuthenticatorCoordinator.cpp:
(WebCore::AuthenticatorCoordinator::create):
(WebCore::AuthenticatorCoordinator::discoverFromExternalSource):
* Source/WebCore/Modules/webauthn/WebAuthenticationUtils.cpp:
(WebCore::buildClientDataJson):
* Source/WebCore/Modules/webauthn/WebAuthenticationUtils.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebAuthenticationPanel.mm:
(produceClientDataJson):
(-[_WKWebAuthenticationPanel makeCredentialWithChallenge:origin:options:completionHandler:]):
(-[_WKWebAuthenticationPanel getAssertionWithChallenge:origin:options:completionHandler:]):
(+[_WKWebAuthenticationPanel getClientDataJSONForAuthenticationType:challenge:origin:]):
(+[_WKWebAuthenticationPanel getClientDataJSONWithTopOrigin:challenge:origin:topOrigin:crossOrigin:]):

Canonical link: <a href="https://commits.webkit.org/263609@main">https://commits.webkit.org/263609@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/54694a07732887a2dba97328748a8f42ddabc41b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/4752 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/4870 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/5033 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/6259 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/4883 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/4745 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/5023 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/4838 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/5117 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/4830 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/4914 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/4252 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/6269 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/2402 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/4245 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/9216 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/4264 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/4319 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/5880 "262 api tests failed or timed out") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/4727 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/3847 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/4241 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1249 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/8296 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/4598 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebCore/Modules/webauthn/WebAuthenticationUtils.cpp:
```diff
@@ -158,7 +158,7 @@ Vector<uint8_t> buildAttestationObject(Vector<uint8_t>&& authData, String&& form
 
 // FIXME(181948): Add token binding ID.
-Ref<ArrayBuffer> buildClientDataJson(ClientDataType type, const BufferSource& challenge, const SecurityOrigin& origin, WebAuthn::Scope scope)
+Ref<ArrayBuffer> buildClientDataJson(ClientDataType type, const BufferSource& challenge, const SecurityOrigin& origin, WebAuthn::Scope scope, std::optional<WTF::String> topOrigin)
```
should the last parameter maybe have a default argument `= std::nullopt` so that you don't have to add it explicitly in the above calls?

cdumez replied:
We almost never use `std::optional<String>` in WebKit because String already has a null state. As a result, you can pass a `const String& topOrigin` here and check `topOrigin.isNull()`.

rr-codes commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebAuthenticationPanel.mm:
```diff
@@ -97,8 +97,10 @@ static inline void updateQueryForGroupIfNecessary(NSMutableDictionary *dictionar
+    auto originType = crossOrigin ? WebAuthn::Scope::CrossOrigin : WebAuthn::Scope::SameOrigin;
 
-    auto clientDataJson = buildClientDataJson(clientDataType, WebCore::BufferSource(challengeBuffer), securityOrigin, WebAuthn::Scope::SameOrigin);
+    auto clientDataJson = buildClientDataJson(clientDataType, WebCore::BufferSource(challengeBuffer), securityOrigin, originType, *topOrigin);
```
what if `topOrigin` is empty here?

abigailfox replied:
Good point, removing *

cdumez commented on Source/WebCore/Modules/webauthn/AuthenticatorCoordinator.cpp:
```diff
@@ -166,7 +166,7 @@ void AuthenticatorCoordinator::create(const Document& document, const PublicKeyC
 
     // Step 14-16.
-    auto clientDataJson = buildClientDataJson(ClientDataType::Create, options.challenge, callerOrigin, scope);
+    auto clientDataJson = buildClientDataJson(ClientDataType::Create, options.challenge, callerOrigin, scope, std::nullopt);
```
{ }

cdumez commented on Source/WebCore/Modules/webauthn/AuthenticatorCoordinator.cpp:
```diff
@@ -241,7 +241,7 @@ void AuthenticatorCoordinator::discoverFromExternalSource(const Document& docume
 
     // Step 10-12.
-    auto clientDataJson = buildClientDataJson(ClientDataType::Get, options.challenge, callerOrigin, scopeAndCrossOriginParent.first);
+    auto clientDataJson = buildClientDataJson(ClientDataType::Get, options.challenge, callerOrigin, scopeAndCrossOriginParent.first, std::nullopt);
```
{ }

cdumez commented on Source/WebCore/Modules/webauthn/WebAuthenticationUtils.cpp:
```diff
@@ -171,6 +171,10 @@ Ref<ArrayBuffer> buildClientDataJson(ClientDataType type, const BufferSource& ch
     object->setString("challenge"_s, base64URLEncodeToString(challenge.data(), challenge.length()));
     object->setString("origin"_s, origin.toRawString());
+    
+    if (topOrigin)
```
!topOrigin.isNull()

cdumez commented on Source/WebCore/Modules/webauthn/WebAuthenticationUtils.cpp:
```diff
@@ -171,6 +171,10 @@ Ref<ArrayBuffer> buildClientDataJson(ClientDataType type, const BufferSource& ch
     object->setString("origin"_s, origin.toRawString());
+    
+    if (topOrigin)
+        object->setString("topOrigin"_s, *topOrigin);
```
topOrigin

cdumez commented on Source/WebCore/Modules/webauthn/WebAuthenticationUtils.h:
```diff
@@ -56,7 +56,7 @@ WEBCORE_EXPORT cbor::CBORValue::MapValue buildCredentialDescriptor(const Vector<
 WEBCORE_EXPORT Vector<uint8_t> buildAttestationObject(Vector<uint8_t>&& authData, String&& format, cbor::CBORValue::MapValue&& statementMap, const AttestationConveyancePreference&);
 
-WEBCORE_EXPORT Ref<ArrayBuffer> buildClientDataJson(ClientDataType /*type*/, const BufferSource& challenge, const SecurityOrigin& /*origin*/, WebAuthn::Scope);
+WEBCORE_EXPORT Ref<ArrayBuffer> buildClientDataJson(ClientDataType /*type*/, const BufferSource& challenge, const SecurityOrigin& /*origin*/, WebAuthn::Scope, std::optional<WTF::String> /*topOrigin*/);
```
no optional.

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebAuthenticationPanel.mm:
```diff
@@ -84,7 +84,7 @@ static inline void updateQueryForGroupIfNecessary(NSMutableDictionary *dictionar
 #endif
 
-static RetainPtr<NSData> produceClientDataJson(_WKWebAuthenticationType type, NSData *challenge, NSString *origin)
+static RetainPtr<NSData> produceClientDataJson(_WKWebAuthenticationType type, NSData *challenge, NSString *origin, std::optional<WTF::String> topOrigin, BOOL crossOrigin)
```
No optional.

cdumez replied:
I recommend changing `BOOL crossOrigin` to `bool isCrossOrigin`. We use a prefix for boolean variables in WebKit and we don't strictly have to use ObjC types here.

cdumez reviewed (commented):
What's up with all the binary files removals?

rr-codes commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebAuthenticationPanel.mm:
```diff
@@ -1050,7 +1052,7 @@ - (void)makeCredentialWithClientDataHash:(NSData *)clientDataHash options:(_WKPu
 {
 #if ENABLE(WEB_AUTHN)
-    auto clientDataJSON = produceClientDataJson(_WKWebAuthenticationTypeGet, challenge, origin);
+    auto clientDataJSON = produceClientDataJson(_WKWebAuthenticationTypeGet, challenge, origin, WTF::nullString(), NO);
```
should probably be `false` now instead of `NO`

cdumez replied:
nullString() -> { }

false instead of NO indeed.

cdumez commented on Source/WebCore/Modules/webauthn/AuthenticatorCoordinator.cpp:
```diff
@@ -166,7 +166,7 @@ void AuthenticatorCoordinator::create(const Document& document, const PublicKeyC
 
     // Step 14-16.
-    auto clientDataJson = buildClientDataJson(ClientDataType::Create, options.challenge, callerOrigin, scope);
+    auto clientDataJson = buildClientDataJson(ClientDataType::Create, options.challenge, callerOrigin, scope, WTF::nullString());
```
just `{ }`

cdumez commented on Source/WebCore/Modules/webauthn/AuthenticatorCoordinator.cpp:
```diff
@@ -241,7 +241,7 @@ void AuthenticatorCoordinator::discoverFromExternalSource(const Document& docume
 
     // Step 10-12.
-    auto clientDataJson = buildClientDataJson(ClientDataType::Get, options.challenge, callerOrigin, scopeAndCrossOriginParent.first);
+    auto clientDataJson = buildClientDataJson(ClientDataType::Get, options.challenge, callerOrigin, scopeAndCrossOriginParent.first, WTF::nullString());
```
just `{ }`

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebAuthenticationPanel.mm:
```diff
@@ -84,7 +84,7 @@ static inline void updateQueryForGroupIfNecessary(NSMutableDictionary *dictionar
 #endif
 
-static RetainPtr<NSData> produceClientDataJson(_WKWebAuthenticationType type, NSData *challenge, NSString *origin)
+static RetainPtr<NSData> produceClientDataJson(_WKWebAuthenticationType type, NSData *challenge, NSString *origin, const String& topOrigin, bool crossOrigin)
```
`isCrossOrigin`

cdumez replied:
The mixing of types here isn't nice, we should be more consistent. origin being a NSString* and topOrigin being a `const String&` is inconsistent, even though they are similar things.

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebAuthenticationPanel.mm:
```diff
@@ -1096,7 +1098,18 @@ + (NSData *)getClientDataJSONForAuthenticationType:(_WKWebAuthenticationType)typ
+    return clientDataJSON.autorelease();
+}
+
++ (NSData *)getClientDataJSONWithTopOrigin:(_WKWebAuthenticationType)type challenge:(NSData *)challenge origin:(NSString *)origin topOrigin:(NSString *)topOrigin crossOrigin:(bool)crossOrigin
```
Any reason the implementation cannot figure out the scope using topOrigin and origin? Why do we need the crossOrigin parameter?

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebAuthenticationPanel.mm:
```diff
@@ -1096,7 +1098,18 @@ + (NSData *)getClientDataJSONForAuthenticationType:(_WKWebAuthenticationType)typ
+    return clientDataJSON.autorelease();
+}
+
++ (NSData *)getClientDataJSONWithTopOrigin:(_WKWebAuthenticationType)type challenge:(NSData *)challenge origin:(NSString *)origin topOrigin:(NSString *)topOrigin crossOrigin:(bool)crossOrigin
```
Any reason the implementation cannot figure out the scope using topOrigin and origin? Why do we need the crossOrigin parameter?

abigailfox replied:
Adding this was requested by client, see radar

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebAuthenticationPanel.mm:
```diff
@@ -1096,7 +1098,18 @@ + (NSData *)getClientDataJSONForAuthenticationType:(_WKWebAuthenticationType)typ
+    return clientDataJSON.autorelease();
+}
+
++ (NSData *)getClientDataJSONWithTopOrigin:(_WKWebAuthenticationType)type challenge:(NSData *)challenge origin:(NSString *)origin topOrigin:(NSString *)topOrigin crossOrigin:(bool)crossOrigin
```
Any reason the implementation cannot figure out the scope using topOrigin and origin? Why do we need the crossOrigin parameter?

abigailfox replied:
Adding this was requested by client, see radar

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebAuthenticationPanel.mm:
```diff
@@ -97,8 +97,10 @@ static inline void updateQueryForGroupIfNecessary(NSMutableDictionary *dictionar
     auto challengeBuffer = ArrayBuffer::tryCreate(reinterpret_cast<const uint8_t*>(challenge.bytes), challenge.length);
     auto securityOrigin = WebCore::SecurityOrigin::createFromString(origin);
+    
+    auto originType = crossOrigin ? WebAuthn::Scope::CrossOrigin : WebAuthn::Scope::SameOrigin;
```
So we don't care about the `WebAuthn::Scope::SameSite` case?

cdumez replied:
It seems buildClientDataJson() only cares about CrossOrigin vs not CrossOrigin so this is fine.,

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebAuthenticationPanel.mm:
```diff
@@ -982,7 +984,7 @@ static _WKAuthenticatorAttachment authenticatorAttachmentToWKAuthenticatorAttach
 {
 #if ENABLE(WEB_AUTHN)
-    auto clientDataJSON = produceClientDataJson(_WKWebAuthenticationTypeCreate, challenge, origin);
+    auto clientDataJSON = produceClientDataJson(_WKWebAuthenticationTypeCreate, challenge, origin, WTF::nullString(), NO);
```
{ }, false

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebAuthenticationPanel.mm:
```diff
@@ -84,7 +84,7 @@ static inline void updateQueryForGroupIfNecessary(NSMutableDictionary *dictionar
 #endif
 
-static RetainPtr<NSData> produceClientDataJson(_WKWebAuthenticationType type, NSData *challenge, NSString *origin)
+static RetainPtr<NSData> produceClientDataJson(_WKWebAuthenticationType type, NSData *challenge, NSString *origin, NSString *topOrigin, bool crossOrigin)
```
I recommend passing a `WebAuthn::Scope` for last parameter since that's what we want in the end. We don't like boolean parameters like this since it makes call sites look confusing when they pass false/true.

Also, can we add default parameter values for the last 2 parameters like somebody already suggested in an earlier review? It would avoid having to update some of the call sites.

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebAuthenticationPanel.mm:
```diff
@@ -982,7 +984,7 @@ static _WKAuthenticatorAttachment authenticatorAttachmentToWKAuthenticatorAttach
 {
 #if ENABLE(WEB_AUTHN)
-    auto clientDataJSON = produceClientDataJson(_WKWebAuthenticationTypeCreate, challenge, origin);
+    auto clientDataJSON = produceClientDataJson(_WKWebAuthenticationTypeCreate, challenge, origin, { }, false);
```
default value for NSString* would read better as `nil` since you switched to NSString. Also, if you used default parameter values, we wouldn't need to update such call sites.

cdumez commented on Source/WebCore/Modules/webauthn/WebAuthenticationUtils.h:
```diff
@@ -56,7 +56,7 @@ WEBCORE_EXPORT cbor::CBORValue::MapValue buildCredentialDescriptor(const Vector<
 WEBCORE_EXPORT Vector<uint8_t> buildAttestationObject(Vector<uint8_t>&& authData, String&& format, cbor::CBORValue::MapValue&& statementMap, const AttestationConveyancePreference&);
 
-WEBCORE_EXPORT Ref<ArrayBuffer> buildClientDataJson(ClientDataType /*type*/, const BufferSource& challenge, const SecurityOrigin& /*origin*/, WebAuthn::Scope);
+WEBCORE_EXPORT Ref<ArrayBuffer> buildClientDataJson(ClientDataType /*type*/, const BufferSource& challenge, const SecurityOrigin& /*origin*/, WebAuthn::Scope, const String& topOrigin = { }/*topOrigin*/);
```
Please drop `/*topOrigin*/`

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebAuthenticationPanel.mm:
```diff
@@ -84,7 +84,7 @@ static inline void updateQueryForGroupIfNecessary(NSMutableDictionary *dictionar
 #endif
 
-static RetainPtr<NSData> produceClientDataJson(_WKWebAuthenticationType type, NSData *challenge, NSString *origin)
+static RetainPtr<NSData> produceClientDataJson(_WKWebAuthenticationType type, NSData *challenge, NSString *origin, NSString *topOrigin = nil, WebAuthn::Scope crossOrigin = WebAuthn::Scope::SameOrigin)
```
Maybe we can rename `crossOrigin` to `scope` for clarity.

abigailfox replied:
This is what the AuthExp team asked for

g-davidson replied:
We don't have strong opinions about the name, but it's implementing the [crossOrigin](https://w3c.github.io/webauthn/#dom-collectedclientdata-crossorigin) field from WebAuthn.

cdumez replied:
@abigailfox This is internal code and it is a scope, it should be named scope. This had nothing to do with the AuthExp team or the API. Use `BOOL crossOrigin` in the API if that's what the AuthExp prefers. I am commenting on internal code, not API.

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebAuthenticationPanel.mm:
```diff
@@ -1102,6 +1102,17 @@ + (NSData *)getClientDataJSONForAuthenticationType:(_WKWebAuthenticationType)typ
     return clientDataJSON.autorelease();
 }
 
++ (NSData *)getClientDataJSONWithTopOrigin:(_WKWebAuthenticationType)type challenge:(NSData *)challenge origin:(NSString *)origin topOrigin:(NSString *)topOrigin crossOrigin:(WebAuthn::Scope)crossOrigin
```
Where are the header changes for this?

Also, we likely want to keep using a `BOOL crossOrigin` in the API since we cannot expose `WebAuthn::Scope` is an internal type that cannot be used in API.

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebAuthenticationPanel.mm:
```diff
@@ -1102,6 +1102,17 @@ + (NSData *)getClientDataJSONForAuthenticationType:(_WKWebAuthenticationType)typ
+    RetainPtr<NSData> clientDataJSON;
+
+#if ENABLE(WEB_AUTHN)
+    clientDataJSON = produceClientDataJson(type, challenge, origin, topOrigin, crossOrigin);
```
Here you'd convert the BOOL into a WebAuthn::Scope.

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebAuthenticationPanel.mm:
```diff
@@ -1102,6 +1102,21 @@ + (NSData *)getClientDataJSONForAuthenticationType:(_WKWebAuthenticationType)typ
+        clientDataJSON = produceClientDataJson(type, challenge, origin, topOrigin, WebAuthn::Scope::CrossOrigin);
+    else
+        clientDataJSON = produceClientDataJson(type, challenge, origin, topOrigin);
+    
```
Extra blank line.

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebAuthenticationPanel.mm:
```diff
@@ -1102,6 +1102,21 @@ + (NSData *)getClientDataJSONForAuthenticationType:(_WKWebAuthenticationType)typ
+    if (crossOrigin)
+        clientDataJSON = produceClientDataJson(type, challenge, origin, topOrigin, WebAuthn::Scope::CrossOrigin);
+    else
+        clientDataJSON = produceClientDataJson(type, challenge, origin, topOrigin);
```
We could do all this on one line:
`clientDataJSON = produceClientDataJson(type, challenge, origin, topOrigin, crossOrigin ? WebAuthn::Scope::CrossOrigin: WebAuthn::Scope::SameOrigin);`

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebAuthenticationPanel.mm:
```diff
@@ -84,7 +84,7 @@ static inline void updateQueryForGroupIfNecessary(NSMutableDictionary *dictionar
 #endif
 
-static RetainPtr<NSData> produceClientDataJson(_WKWebAuthenticationType type, NSData *challenge, NSString *origin)
+static RetainPtr<NSData> produceClientDataJson(_WKWebAuthenticationType type, NSData *challenge, NSString *origin, NSString *topOrigin = nil, WebAuthn::Scope crossOrigin = WebAuthn::Scope::SameOrigin)
```
Should be named `scope`, not `crossOrigin`.

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebAuthenticationPanel.mm:
```diff
@@ -1102,6 +1102,17 @@ + (NSData *)getClientDataJSONForAuthenticationType:(_WKWebAuthenticationType)typ
     return clientDataJSON.autorelease();
 }
 
++ (NSData *)getClientDataJSONWithTopOrigin:(_WKWebAuthenticationType)type challenge:(NSData *)challenge origin:(NSString *)origin topOrigin:(NSString *)topOrigin crossOrigin:(BOOL)crossOrigin
```
I commented on the earlier iteration but aren't you missing the corresponding declaration in the header?

cdumez reviewed (approved):
LGTM

abigailfox committed:
Add support for topOrigin and crossOrigin to [_WKWebAuthenticationPanel getClientDataJSONFor...]
https://bugs.webkit.org/show_bug.cgi?id=255680
rdar://108064670

Reviewed by J Pascoe and Chris Dumez.

* Source/WebCore/Modules/webauthn/AuthenticatorCoordinator.cpp:
(WebCore::AuthenticatorCoordinator::create):
(WebCore::AuthenticatorCoordinator::discoverFromExternalSource):
* Source/WebCore/Modules/webauthn/WebAuthenticationUtils.cpp:
(WebCore::buildClientDataJson):
* Source/WebCore/Modules/webauthn/WebAuthenticationUtils.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebAuthenticationPanel.mm:
(produceClientDataJson):
(-[_WKWebAuthenticationPanel makeCredentialWithChallenge:origin:options:completionHandler:]):
(-[_WKWebAuthenticationPanel getAssertionWithChallenge:origin:options:completionHandler:]):
(+[_WKWebAuthenticationPanel getClientDataJSONForAuthenticationType:challenge:origin:]):
(+[_WKWebAuthenticationPanel getClientDataJSONWithTopOrigin:challenge:origin:topOrigin:crossOrigin:]):

Canonical link: https://commits.webkit.org/263609@main
